File: src/namespaces/das/cnft-to-das.ts
// Map a cNFT LeafRecord to the DAS asset shape Helius returns. Pure —
// no I/O — so it's trivially testable. The only Kit dependency is the
// Address decoder for base58 encoding of the 32-byte fields we carry
// as Uint8Array internally.

import { getAddressDecoder, getBase58Decoder } from "@solana/kit";
import type { DasAsset } from "../../decoders/index.js";
import type { LeafRecord } from "../../cnft/index.js";

const addrDecoder = getAddressDecoder();
const base58 = getBase58Decoder();

function addressFromBytes(bytes: Uint8Array): string {
  return addrDecoder.decode(bytes) as string;
}

function toBase58(bytes: Uint8Array): string {
  return base58.decode(bytes);
}

export function leafRecordToDasAsset(record: LeafRecord): DasAsset {
  const m = record.mintMetadata;

  const creators = m.creators.map((c) => ({
    address: addressFromBytes(c.address),
    share: c.share,
    verified: c.verified,
  }));

  const grouping: DasAsset["grouping"] = m.collection
    ? [
        {
          group_key: "collection",
          group_value: addressFromBytes(m.collection.key),
        },
      ]
    : [];

  const owner = addressFromBytes(record.owner);
  const delegate = addressFromBytes(record.delegate);
  const delegated = owner !== delegate;

  return {
    id: record.assetId as string,
    interface: "V1_NFT",
    content: {
      $schema: "https://schema.metaplex.com/nft1.0.json",
      json_uri: m.uri,
      metadata: {
        name: m.name,
        symbol: m.symbol,
        description: "",
      },
      links: {
        image: null,
        animation_url: null,
      },
      files: [],
    },
    authorities: [],
    creators,
    ownership: {
      frozen: false,
      delegated,
      ownership_model: "single",
      owner,
    },
    grouping,
    mutable: m.isMutable,
    burnt: record.burned,
    compression: {
      eligible: true,
      compressed: true,
      data_hash: toBase58(record.dataHash),
      creator_hash: toBase58(record.creatorHash),
      asset_hash: toBase58(record.leafHash),
      tree: record.tree as string,
      seq: 0,
      leaf_id: Number(record.leafIndex),
    },
  };
}

File: crates/service/src/das/cnft_to_das.rs
//! Pure mapper: `LeafRecord` (our stored cNFT shape) → `DasAsset` (the
//! external wire format Helius clients expect). No async, no I/O —
//! trivially testable and trivially Rust-portable.

use crate::cnft::LeafRecord;

use super::types::{
    DasAsset, DasAuthority, DasCompression, DasContent, DasCreator, DasFile, DasGrouping,
    DasLinks, DasMetadata, DasOwnership,
};

/// Convert a cNFT `LeafRecord` into the DAS `getAsset` response shape.
#[must_use]
pub fn leaf_record_to_das_asset(record: &LeafRecord) -> DasAsset {
    let m = &record.mint_metadata;

    let creators = m
        .creators
        .iter()
        .map(|c| DasCreator {
            address: bs58::encode(c.address).into_string(),
            share: c.share,
            verified: c.verified,
        })
        .collect();

    let grouping = m
        .collection
        .as_ref()
        .map(|(key, _)| {
            vec![DasGrouping {
                group_key: "collection".into(),
                group_value: bs58::encode(key).into_string(),
            }]
        })
        .unwrap_or_default();

    let owner = bs58::encode(record.owner).into_string();
    let delegate = bs58::encode(record.delegate).into_string();
    let delegated = owner != delegate;

    DasAsset {
        id: bs58::encode(record.asset_id).into_string(),
        interface: "V1_NFT".into(),
        content: DasContent {
            schema: "https://schema.metaplex.com/nft1.0.json".into(),
            json_uri: m.uri.clone(),
            metadata: DasMetadata {
                name: m.name.clone(),
                symbol: m.symbol.clone(),
                description: String::new(),
                ..Default::default()
            },
            links: DasLinks::default(),
            files: Vec::<DasFile>::new(),
            category: None,
        },
        authorities: Vec::<DasAuthority>::new(),
        creators,
        ownership: DasOwnership {
            delegated,
            ownership_model: "single".into(),
            owner,
            ..Default::default()
        },
        grouping,
        mutable: m.is_mutable,
        burnt: record.burned,
        compression: Some(DasCompression {
            eligible: true,
            compressed: true,
            data_hash: bs58::encode(record.data_hash).into_string(),
            creator_hash: bs58::encode(record.creator_hash).into_string(),
            asset_hash: bs58::encode(record.leaf_hash).into_string(),
            tree: bs58::encode(record.tree).into_string(),
            seq: 0,
            leaf_id: record.leaf_index,
        }),
        ..Default::default()
    }
}
